sympy/core/basic.py

```diff
@@ -1944,6 +1944,42 @@ def could_extract_minus_sign(self):
         return False  # see Expr.could_extract_minus_sign
 
 
+class FuncHead():
+    """Acts as the head of the object when its type would not be
+    suitable for the arg invariance obj.func(*obj.args)==obj. An
+    example of this would be Atoms"""
+
+    __slots__ = ('obj')
+    def __new__(cls, obj):
+        o = object.__new__(cls)
+        o.obj = obj
+        return o
+
+    def __instancecheck__(self, other):
+        if isinstance(other, type(self.obj)):
+            return True
+        return super().__instancecheck__(other)
+
+    def __call__(self, *args):
+        if args:
+            return type(self.obj)(*args)
+        else:
+            return self.obj
+
+    def __eq__(self, other):
+        if self.obj.__class__ == other: return True
+        return super().__eq__(other)
+
+    def __hash__(self):
+        return hash(self.obj.__class__)
+
+    def __repr__(self):
+        return repr(self.obj.__class__)
```

Comment:
Probably best to print as something like `_FuncHead(self.obj.__class__)` or something else that just makes it clear that this isn't actually the type. It's a weird object so it should look like one.